### PR TITLE
Add new gloo.js with an API

### PR DESCRIPTION
**Pull request by almarklein**
See issue #1 

This adds a gloo2.js, providing a gloo JS API, on top of which we can write the GLIR parser. The code currently does not support FBO's. 

The JS code is created by taking the glir.py, and modifying it slightly so it is more tuned for webgl, accepted by PyScript, which translates the code to JS. I'm not sure if this is the best way to do it, but it seemed the fastest way to get the API up, and it works! We could decide to just take the gloo.js from here and ditch the Python version.

Right now, to compile the gloo2.js, you need to install Flexx (https://github.com/zoofio/flexx), which is in itself very much a WIP, but the PyScript module is quite complete and well-tested.

### CODE REVIEW

File: scripts/gloo2.js [hunk 1]
```diff
@@ -0,0 +1,560 @@
+/* Do not edit, autogenerated by flexx.pyscript */
+
```
Comment (almarklein) [inline scripts/gloo2.js:2]: Probably also do not review; review the .py instead.

File: scripts/gloo2.py [hunk 2]
```diff
@@ -0,0 +1,475 @@
...
+                return
+        # Look up function to call
+        funcname = self.UTYPEMAP[type_]
+        # Program needs to be active in order to set uniforms
+        self.activate()
+        self._gl[funcname](handle, value)
+    
+    # todo: Better api
+    def set_attribute(self, name, type_, value):
+        """ Set an attribute value. Value is assumed to have been checked.
```
Comment (almarklein) [inline scripts/gloo2.py]: I wrote that todo some time ago. Not sure what I had in mind. Maybe better docs should be sufficient. Talking of docs, what is a sensible way to document the gloo API. Should we put it on RTD?

File: scripts/gloo2.py [hunk 3]
```diff
@@ -0,0 +1,475 @@
...
+    
+    # todo: is this not always 32bit? for attributes yes...
+    def set_size(self, nbytes):  # in bytes
+        if nbytes != self._buffer_size:
+            self.activate()
+            self._gl.bufferData(self._target, nbytes, self._usage)
+            self._buffer_size = nbytes
+    
+    #  todo: make possible to also resize here, first init, flag?
+    def set_data(self, offset, data):
```
Comment (almarklein) [inline scripts/gloo2.py:372]: It gets a bit tedious to always have to set both size and data. Options:
- A: Keep it; tedious but clear
- B: set the size automatically the first time that the data is set (init `_buffer_size` with None) 
- C: extra flag to `set_data(self, offset, data, set_size=False)`

File: scripts/gloo2.py [hunk 4]
```diff
@@ -0,0 +1,474 @@
...
+            self._gl.pixelStorei(self._gl.UNPACK_ALIGNMENT, 4)
+
+
+if __name__ == '__main__':
+    import os
+    from flexx.pyscript.functions import script2js
+    script2js(__file__, 'gloo2')
+    
+    # AK: small hack to easily deploy for Bokeh
+    bokehdir = '/home/almar/dev/pylib/bokeh/bokehjs/src/vendor/gloo/'
```
Comment (rossant) [inline scripts/gloo2.py:479]: this should not be part of the repo..?

**Comment by rossant** [comment]
the approach looks interesting. I'd be in favor of _not_ stitching the python version, but rather keep it and  work on it. I even think that the auto-generated gloo2.js should not be in the repo: it could be generated automatically during the build process. I find it dangerous to have both a source and autogenerated target  in the same repo...

also, do you plan to work on FBO in this PR (or a future PR)? same question for the glir interpreter?

finally, is it conceivable to use exactly the same gloo2.py in the main vispy branch?

File: scripts/gloo2.py (hunk 4 above)
Comment (almarklein) [inline scripts/gloo2.py:479]: not really, but it saves me some time. I suppose I could remove it :)

**Comment by almarklein** [comment]
>  I find it dangerous to have both a source and autogenerated target in the same repo...

there's a message at the top: _Do not edit, autogenerated by flexx.pyscript_  although this is not always foolproof...

> do you plan to work on FBO in this PR (or a future PR)? 

It should not cost too much time to include that too. But I am not sure how to test it; we'd need a JS example that uses FBO's, or we test it after writing the glir interpreter on top ...

> same question for the glir interpreter?

I would not mind if someone else did that; I have no use for it right now. Should be straightforward.

> is it conceivable to use exactly the same gloo2.py in the main vispy branch?

I considered this (how cool would that be!) but I changed the code to take subtle webgl stuff into account.

**Comment by almarklein** [comment]
> > is it conceivable to use exactly the same gloo2.py in the main vispy branch?
> > I considered this (how cool would that be!) but I changed the code to take subtle webgl stuff into account.

Mmm, if we put some if-statements in the code and had some `ISJS` constant, it might work ...  One problem: here we use `gl.xxx`, while in Python its `gl.glXxx`. If we can get around that, we might get it to work!

**Pull request merged**